### PR TITLE
 Commuter Patterns: Larger shapes for border intersections

### LINKED ISSUE #215: Render border intersections better
From conversations with Michael and Yuwen, it's becoming apparent that intersections along the map borders aren't obvious. Especially in the commuter map, a huge percent of trips might be distributed to borders, but they're drawn way too small to notice. Strawman idea is to attach much larger arrows (inbound, outbound, or both) to the borders. At least for the commuter map, but maybe always -- this could be the new shape for them in general.

Screenshot from 2020-07-20 13-50-21
Screenshot from 2020-07-20 13-48-03

### PR DESCRIPTION
**Pull request by michaelkirk**
Fixes #215 

A substantial portion of traffic has an origin/destination off the map.
We were correctly tracking all this data, attributing it to a
"block" the size of the border "intersection"; However, the design
didn't accommodate the small "size" of the border intersections:

**before**

image

1. it was too small to meaningfully reason about the heatmap visualization
2. it was too small of a "hit" area for clicking and hovering
3. when in the "locked" mode, the hover text overlay was illegible.

**after**

Now we draw a larger arrow, representing the inward/outward/both-ways
direction of traffic which makes it easier to reason about the heatmap
color and accommodates an adequate hit area.

And the hover text is now printed over a white pill to maintain
legibility.

Screen Shot 2020-07-27 at 12 16 55 PM

demo mov

### CODE REVIEW

File: game/src/sandbox/dashboards/commuter.rs [hunk 1]
```diff
@@ -211,6 +211,15 @@ impl CommuterPatterns {
 
                 batch.push(Color::BLACK.alpha(0.5), base_block.shape.clone());
 
+                // Draw outline for Locked Selection
```
Comment (michaelkirk) [inline game/src/sandbox/dashboards/commuter.rs:214]: existing issue that became obvious with this PR, we should draw the border beneath the inwar/outward icon, not above it.

File: game/src/sandbox/dashboards/commuter.rs [hunk 2]
```diff
@@ -274,12 +274,16 @@ impl CommuterPatterns {
...
-                                .centered_on(compare_to_block.shape.polylabel()),
-                        );
+                        let label = Text::from(Line(label_text).fg(Color::BLACK))
+                            .render_to_batch(ctx.prerender)
+                            .scale(2.0)
+                            .centered_on(compare_to_block.shape.polylabel());
+
+                        let dims = label.get_dims();
+                        let label_bg =  Polygon::rounded_rectangle(dims.width + 70.0, dims.height + 20.0, None);
+                        let bg = GeomBatch::from(vec![(Color::WHITE, label_bg)]).centered_on(compare_to_block.shape.polylabel());
```
Comment (michaelkirk) [inline game/src/sandbox/dashboards/commuter.rs:284]: I'm kind of ambivalent about the pill shaped background - but we've got to do *something* to make the text legible. WDYT?

Comment (dabreegster) [reply game/src/sandbox/dashboards/commuter.rs:284]: I personally like it, even on nice large blocks. Even easier to see.

File: game/src/sandbox/dashboards/commuter.rs [hunk 3]
```diff
@@ -498,26 +502,71 @@ fn group_bldgs(
...
             borders: hashset! { i.id },
-            shape: i.polygon.clone(),
+            shape: build_shape_for_border(i, BorderType::Outgoing, &app.primary.map),
         });
     }
 
     (bldg_to_block, border_to_block, blocks)
 }
 
+enum BorderType {
```
Comment (dabreegster) [inline game/src/sandbox/dashboards/commuter.rs:528]: This is nice, I might lift it to `map_model` later.

File: game/src/sandbox/dashboards/commuter.rs [hunk 4]
```diff
@@ -498,26 +502,71 @@ fn group_bldgs(
...
+}
+
+fn build_shape_for_border(
+    border: &map_model::Intersection,
+    border_type: BorderType,
+    map: &Map,
+) -> Polygon {
+    let start = border.polygon.center();
+
+    let road = match border_type {
```
Comment (dabreegster) [inline game/src/sandbox/dashboards/commuter.rs:541]: By definition, a border only has one road. You could just do `map.get_r(*border.roads.iter().next().unwrap())`. Although it's sort of gross to dig out an element guaranteed to exist in a `BTreeSet`.

File: geom/src/polyline.rs [hunk 5]
```diff
@@ -543,6 +543,56 @@ impl PolyLine {
         Polygon::new(&pts)
     }
 
+    pub fn make_double_arrow(&self, thickness: Distance, cap: ArrowCap) -> Polygon {
```
Comment (dabreegster) [inline geom/src/polyline.rs:546]: Nice!

**Comment by dabreegster** [comment]
Two tiny fixes, then we can merge this. Thanks! Huge improvement. It's worth considering drawing these arrows all of the time, though I might tackle https://github.com/dabreegster/abstreet/issues/220 first, which could change how they'd look.

File: game/src/sandbox/dashboards/commuter.rs [hunk 6]
```diff
@@ -498,26 +502,71 @@ fn group_bldgs(
...
+    } else {
+        center_line.first_line().angle()
+    };
+
+    let length = Distance::meters(150.0);
+    let thickness = Distance::meters(30.0);
+    let end = start.project_away(length, angle);
+
+    match border_type {
+        BorderType::Incoming => PolyLine::new(vec![end, start])
```
[the remark this answers is not in the document]
Comment (michaelkirk) [reply game/src/sandbox/dashboards/commuter.rs]: done!

File: game/src/sandbox/dashboards/commuter.rs (hunk 4 above)
Comment (michaelkirk) [reply game/src/sandbox/dashboards/commuter.rs]: done!

**Comment by michaelkirk** [comment]
Ok, I think I've responded to all your feedback. Please take another look @dabreegster

**Pull request merged**